source/client/service_impl.cc

```diff
@@ -66,13 +67,18 @@ ::grpc::Status ServiceImpl::ExecutionStream(
                                ::nighthawk::client::ExecutionRequest>* stream) {
   nighthawk::client::ExecutionRequest request;
   stream_ = stream;
+  busy_ = false;
   while (stream->Read(&request)) {
     ENVOY_LOG(debug, "Read ExecutionRequest data: {}", request.DebugString());
     if (request.has_start_request()) {
-      if (future_.valid() &&
-          future_.wait_for(std::chrono::seconds(0)) != std::future_status::ready) {
+      // It is possible to receive a back-to-back request here, while the future that is associated
+      // to our previous response is still active. We check the running_ flag to see if the previous
+      // future has progressed in a state where we can do another one. This avoids the odd flake in
+      // ServiceTest.BackToBackExecution.
```

Comment:
I'm trying to figure out how this works.. it seems that if we had a previous request with an active future, won't we be still waiting on it in the previous requst's `finishGrpcStream`?

Reply:
Well, we're testing servicing of back-to-back execution on the same stream in this test, `finishGrpcStream` isn't involved in the failure. The failure happens when the client queues up the second request on the same stream right after receiving a response to the first one. The future that was responsible for servicing the reply can still be working to tear down. We only want to ensure that we have a single process context active at any time, so the proposed fix here is to indicate we are ready for the next request from within the code associated to the future (right before putting the reply on the stream, but after tearing down the process context).

Reply:
Can we just do a zero timeout with https://en.cppreference.com/w/cpp/thread/future/wait_for to check future state?

Thinking about future semantics though, are we sure that the future even begins execution until we do a wait on it? An efficient runtime should do this, but I wonder if we're making some non-standard guaranteed assumption here.

Reply:
Done, In the change I pushed we:
- wait with a zero timeout upon observing an active future. when the timeout fails we check the new busy flag.
- we start explicitly passing `std::launch::async`, to avoid running into lazy execution, good point